cores/esp8266/Esp.cpp

```diff
@@ -673,7 +673,48 @@ bool EspClass::flashEraseSector(uint32_t sector) {
     return rc == 0;
 }
 
+// Adapted from the old version of `flash_hal_write()` (before 3.0.0), which was used for SPIFFS to allow
+// writing from both unaligned u8 buffers and to an unaligned offset on flash.
+// Updated version re-uses some of the code from RTOS, replacing individual methods for block & page
+// writes with just a single one
+// https://github.com/espressif/ESP8266_RTOS_SDK/blob/master/components/spi_flash/src/spi_flash.c
+
+// Wrapper around spi_flash_write, handling offset + size crossing page boundaries
+// Note that we expect that both `offset` and `data` are properly aligned
+static SpiFlashOpResult spi_flash_write_page_break(uint32_t offset, uint32_t* data, size_t size) {
+    static constexpr uint32_t PageSize { FLASH_PAGE_SIZE };
+    size_t page_size = PageSize - (offset % PageSize);
+
+    // most common case, we don't cross a page and simply write the data
+    if (size < page_size) {
+        return spi_flash_write(offset, data, size);
+    }
+
+    // otherwise, write the initial part and continue writing breaking each page interval
+    SpiFlashOpResult result = SPI_FLASH_RESULT_ERR;
+    if ((result = spi_flash_write(offset, data, page_size)) != SPI_FLASH_RESULT_OK) {
+        return result;
+    }
+
+    uint32_t page_offset = PageSize;
+    for (uint32_t page = 0; page < (size - page_size) / PageSize; ++page) {
```

Comment:
While you're at it optimizing speed, better move the `(size - page_size) / PageSize` part of the for loop call into a const declared variable before calling the for loop.